File: 0649-dota2-senate/0649-dota2-senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        //create two queues to keep track of eligible senators separately in sorted order
        int n = senate.size();
        queue<int> rQueue;
        queue<int>dQueue;

        //populate queues with indicies of respective senators from left to right
        for (int i = 0; i < n; i++) {
            if (senate[i] == 'R') {
                rQueue.push(i);
            } else {
                dQueue.push(i);
            }
        }

        //while both Queues have at least one senator
        while (!rQueue.empty() && !dQueue.empty()) {
            //pop Next-Turn Sentate from both Queues
            int rTurn = rQueue.front();
            rQueue.pop();
            int dTurn = dQueue.front();
            dQueue.pop();

            //larger index is banned. Smaller index en-Queue + N for next round turn
            if (dTurn < rTurn) {
                dQueue.push(dTurn + n);
            } else {
                rQueue.push(rTurn + n);
            }
        }

        //return non empty
        return rQueue.empty() ? "Dire" : "Radiant";
    } //runtime : O(N), space : O(N)
};

File: 0649-dota2-senate/0649-dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        n = len(senate)
        
        #fill queues with senators index
        r_queue = deque()
        d_queue = deque()
        
        #populate queues
        for i,s in enumerate(senate):
            if s == 'R':
                r_queue.append(i)
            else:
                d_queue.append(i)
                
        #while both parties have senator
        while r_queue and d_queue:
            #pop next turn from both and compare
            r_turn = r_queue.popleft()
            d_turn = d_queue.popleft()
            #add leftmost back into queue
            if d_turn < r_turn:
                d_queue.append(d_turn + n)
            else:
                r_queue.append(r_turn + n)
                
        return "Radiant" if r_queue else "Dire"
        